test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs

```diff
@@ -35,14 +35,12 @@ protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage
                 string state = queryString["state"];
 
                 queryString.Clear();
-
+                var builder = new UriBuilder(location);
```

Comment:
This looks like you're undoing what the code was trying to achieve, which was stripping out all the query string parameters except those explicitly in use (`code` and `state`). If you need to retain `redirect_uri` then you just need to read it out into a local like on line 38 and then put it back like on line 42.

Reply:
But this is not a general parameter. I put state in redirect_uri to solve the problem of too long state parameter.

Reply:
From what I can see in the diff, you're forwarding on everything in the original query string for the request to the OAuth provider back to the app under test.

The original code is only forwarding back the parameters that were needed, which are `code` and `state`. If you also need `redirect_uri`, then just add that as special case too.

If we were going to just reflect everything back, then the majority of the code in that method becomes redundant and could be simplified compared to what's there now in your change as all the special-case code would become defunct as it would be passed back anyway.

The reason to only pick-and-choose what's reflected back is to remove noise from provider-specific parameters in the call back to the test app as part of the loopback is just to make it easier to debug.

Reply:
so?
                var builder = new UriBuilder(location);
                // Retain the parameters of redirect_uri
                queryString = HttpUtility.ParseQueryString(builder.Query);
                var oauthstate= queryString["_oauthstate"];
                queryString.Clear();

                queryString.Add("code", "a6ed8e7f-471f-44f1-903b-65946475f351");
                queryString.Add("state", state);
                queryString.Add("_oauthstate", oauthstate);
                builder.Query = queryString.ToString();

Reply:
So like the below, you should only need to add two lines of code.

```csharp
string state = queryString["state"];
string oauthState = queryString["_oauthstate"];

queryString.Clear();

queryString.Add("code", "a6ed8e7f-471f-44f1-903b-65946475f351");
queryString.Add("state", state);
queryString.Add("_oauthstate", oauthState);

var builder = new UriBuilder(location)
{
    Query = queryString.ToString(),
};
```

Reply:
_ oauthstate is in redirect_uri, so it needs to be parsed from location-_-

Reply:
Right I see, so then you only need to pass back `redirect_uri` then, not `_oauthstate`?

Reply:
Yes

Reply:
You need to take `_oauthstate `from `redirect_uri `and resolve it to `state`
Look at the code
```
protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync()
        {
            if (!IsWeixinAuthorizationEndpointInUse())
            {
                if (Request.Query.TryGetValue(OauthState, out var stateValue))
                {
                    var query = Request.Query.ToDictionary(c => c.Key, c => c.Value, StringComparer.OrdinalIgnoreCase);
                    if (query.TryGetValue(State, out var _))
                    {
                        query[State] = stateValue;
                        Request.QueryString = QueryString.Create(query);
                    }
                }
            }
            return await base.HandleRemoteAuthenticateAsync();
        }
```

Reply:
So for the tests you just need `redirect_uri` passed back then.

Reply:
`code` is also required, so please check my submission

Reply:
code is also required, so please check my submission

Reply:
So what should I do now?

Reply:
I'll have time to re-review this later today, or at the latest this weekend.